File: cpp_drill/Problems/STK02_MinStack.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

class MinStack {
private:
    stack<int> st;
    stack<int> min_st;
public:
    MinStack() {};
    
    void push(int val) {
        st.push(val);
        int min_val = val;
        if (!min_st.empty())
            min_val = min_st.top();
        val = min(val, min_val);
        min_st.push(val);
    }
    
    void pop() {
        st.pop();
        min_st.pop();
    }
    
    int top() {
        return st.top();
    }
    
    int getMin() {
        return min_st.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */


int main() {
    return 0;
}

File: py_drill/Problems/STK02_MinStack.py
# REF : https://leetcode.com/problems/min-stack/
# NOTES :
#       Keep two stack one for actual stack and another for min value so far
#       length of both should be same

# T: O(1)
# S: O(n)

class MinStack:

    def __init__(self):
        self.stack = []
        self.minStack = []

    def push(self, val: int) -> None:
        self.stack.append(val)

        min_val = val
        if self.minStack:
            min_val = self.minStack[-1]

        val = min(val, min_val)
        # val = min(val, self.minStack[-1] if self.minStack else val)
        self.minStack.append(val)

    def pop(self) -> None:
        self.stack.pop()
        self.minStack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.minStack[-1]



# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

def main():
    pass

if __name__ == "__main__":
    main()